ppafm/ocl/field.py

```diff
@@ -277,6 +277,70 @@ def to_file(self, file_path, clamp=None):
             io.save_vec_field(file_head, array[:, :, :, :3], self.lvec, data_format=ext)
             io.save_scal_field(file_head + "_w", array[:, :, :, 3], self.lvec, data_format=ext)
 
+    def _prepare_same_size_output_grid(self, array_in, in_place):
+        if in_place:
+            grid_out = self
+            self._array = None  # The current host array will be wrong after operation, so reset it
+        else:
+            array_out = cl.Buffer(self.ctx, cl.mem_flags.READ_WRITE, size=array_in.size)
+            array_type = type(self)  # This way so inherited classes return their own class type
+            grid_out = array_type(array_out, lvec=self.lvec, shape=self.shape, ctx=self.ctx)
+        return grid_out
+
+    def clamp(self, minimum=-np.inf, maximum=np.inf, clamp_type="hard", soft_clamp_width=1.0, in_place=True, local_size=(32,), queue=None):
+        """
+        Clamp data grid values to a specified range. The ``'hard'`` clamp simply clips values that are out of range,
+        and the ``'soft'`` clamp uses a sigmoid to smoothen the transition.
+
+        Arguments:
+            minimum: float. Values below minimum are set to minimum.
+            maximum: float. Values above maximum are set to maximum.
+            clamp_type: str. Type of clamp to use: ``'soft'`` or ``'hard'``.
+            soft_clamp_width: float. Width of transition region for soft clamp.
+            in_place: bool. Whether to do operation in place or to create a new array.
+            local_size: tuple of a single int. Size of local work group on device.
+            queue: pyopencl.CommandQueue. OpenCL queue on which operation is performed. Defaults to oclu.queue.
+
+        Returns:
+            grid_out: Same type as self. New data grid with result.
+        """
+
+        array_in = self.cl_array
+        grid_out = self._prepare_same_size_output_grid(array_in, in_place)
+        n = np.int32(array_in.size / 4)
+        minimum = np.float32(minimum)
+        maximum = np.float32(maximum)
+        soft_clamp_width = np.float32(soft_clamp_width)
+
+        queue = queue or oclu.queue
+        global_size = [int(np.ceil(n / local_size[0]) * local_size[0])]
+
+        if clamp_type == "hard":
+            # fmt: off
+            cl_program.clamp_hard(queue, global_size, local_size,
+                array_in,
+                grid_out.cl_array,
+                n,
+                minimum,
+                maximum,
+            )
+            # fmt: on
+        elif clamp_type == "soft":
```

Comment:
nice possibility of the soft clamping - is it useable through CLI?

Reply:
No, I only implemented here for the GPU when I was testing different clamping types. I decided that the soft clamp is not really necessary, so I didn't bother putting it elsewhere. Do you think this would be useful as a function in the CPU code?

Reply:
Ok, this is something we should talk about in the meeting. As in principle, I like to have a lot of possibilities, there is already a lot of code unused, that is why.
Good for now,